rust/quic/src/quic.rs

```diff
@@ -0,0 +1,263 @@
+// written by Victor Julien
+// partly written by Pierre Chifflier
+extern crate libc;
+use std;
+use std::mem;
+use libc::c_char;
+//use std::ffi::CStr;
+
+use nom::{rest};
+use nom::IResult;
+use rparser::*;
+
+use std::io::Write;
+
+pub fn bytes_to_u64(s: &[u8]) -> Result<u64, &'static str> {
+    let mut u = 0;
+
+    for &c in s {
+        u *= 256;
+        u += c as u64;
+    }
+
+    Ok(u)
+}
+
+macro_rules! println_stderr(
+    ($($arg:tt)*) => { {
+        let r = writeln!(&mut ::std::io::stderr(), $($arg)*);
+        r.expect("failed printing to stderr");
+    } }
+);
+
+pub struct QuicParserState<'a> {
+    pub o: Option<&'a[u8]>,
+    events: Vec<u32>,
+    counter: u64,
+}
+
+impl<'a> QuicParserState<'a> {
+    pub fn new(i: &'a[u8]) -> QuicParserState<'a> {
+        QuicParserState {
+            o:Some(i),
+            events:Vec::new(),
+            counter:0,
+        }
+    }
+}
+
+r_declare_state_new!(r_quic_state_new,QuicParserState,b"blah");
+r_declare_state_free!(r_quic_state_free,QuicParserState,{});
+
+#[derive(Debug,PartialEq)]
+pub struct QuicPacket<'a> {
+    pub flags_has_version: bool,
+    pub flags_has_reset: bool,
+    pub conn_fld_len: u8,
+    pub connection_id: Option<&'a[u8]>,
+    pub version: Option<&'a[u8]>,
+    pub seq_fld_len: u8,
+    pub seq_id: &'a[u8],
+    pub payload: &'a[u8],
+}
+
+named!(pub parse_quic<QuicPacket>,
+   do_parse!(
+       flags: bits!(tuple!(
+               take_bits!(u8, 2),   // reserved
+               take_bits!(u8, 2),   // seq
+               take_bits!(u8, 2),   // conn,
+               take_bits!(u8, 1),   // reset
+               take_bits!(u8, 1)))  // has ver
+
+       // get connection field length and take it
+       >> conn_len: value!(match flags.2 { 0x00 => 0, 0x01 => 1, 0x02 => 4, /* 0x03 */_ => 8, })
+       >> connid: cond!(flags.2 > 0, take!(conn_len))
+
+       // if version is present it's always 4 bytes
+       >> ver : cond!(flags.4 == 1, take!(4))
+
+       // sequence is 1, 2, 4 or 6 bytes
+       >> seq_len: value!(match flags.1 { 0x00 => 1, 0x10 => 2, 0x20 => 4, /* 0x30 */_ => 6, })
+       >> seqid: take!(seq_len)
+
+       >> pl: rest
+
+       >> (
+           QuicPacket {
+               flags_has_version:flags.4 == 1,
+               flags_has_reset:flags.3 == 1,
+               conn_fld_len:conn_len,
+               version:ver,
+               seq_fld_len:seq_len,
+               connection_id:connid,
+               seq_id:seqid,
+               payload: pl,
+           }
+   ))
+);
+
+struct QuicParser;
+
+impl<'a> RParser<QuicParserState<'a>> for QuicParser {
+    fn new_state() -> QuicParserState<'a> {
+        QuicParserState::new(b"blah")
+    }
+
+    fn probe(i: &[u8]) -> bool {
+        match parse_quic(i) {
+            IResult::Done(_, quic_packet) => {
```

Comment:
If you don't use `quic_packet` you can use `IResult::Done(_,_)` to avoid an unused variable warning